pennylane/tape/tape.py

```diff
@@ -15,10 +15,11 @@
 This module contains the base quantum tape.
 """
 # pylint: disable=too-many-instance-attributes,protected-access,too-many-branches,too-many-public-methods
-from collections import Counter, deque
+from collections import Counter, deque, defaultdict
```

Comment:
curious what defaultdict is?

Reply:
If the key doesn't exist, it will have some default value, either 0 for `int` or `[]` for `list`.  

This way we don't have to write :
```
dictionary[key] = added + dictionary.get("key", 0)
```

Also, users can get `0` for objects that aren't in the dictionary.  Easier to determine that there aren't any gates of a particular type of name.